Source/playerdat.cpp

```diff
@@ -152,11 +155,122 @@ void ReloadExperienceData()
 	}
 }
 
+void LoadClassAttributes(std::string_view classPath, ClassAttributes &out)
+{
+	const std::string filename = StrCat("txtdata\\classes\\", classPath, "\\attributes.tsv");
+	tl::expected<DataFile, DataFile::Error> dataFileResult = DataFile::load(filename);
+	if (!dataFileResult.has_value()) {
+		DataFile::reportFatalError(dataFileResult.error(), filename);
+	}
+	DataFile &dataFile = dataFileResult.value();
+
+	enum class ClassAttributeColumn {
+		Attribute,
+		Value,
+		LAST = Value
+	};
+	std::array<ColumnDefinition, enum_size<ClassAttributeColumn>::value> columns;
+	if (tl::expected<void, DataFile::Error> result
+	    = dataFile.parseHeader<ClassAttributeColumn>(columns.data(), columns.data() + columns.size(),
+	        [](std::string_view name) -> tl::expected<ClassAttributeColumn, ColumnDefinition::Error> {
+		        if (name == "Attribute")
+			        return ClassAttributeColumn::Attribute;
+		        if (name == "Value")
+			        return ClassAttributeColumn::Value;
+		        return tl::make_unexpected(ColumnDefinition::Error::UnknownColumn);
+	        });
+	    !result.has_value()) {
+		DataFile::reportFatalError(result.error(), filename);
+	}
+
+	struct KeyInfo {
+		std::variant<uint8_t *, int16_t *> out;
+		bool found = false;
+	};
+	std::unordered_map<std::string_view, KeyInfo> keyToField {
+		{ "baseStr", KeyInfo { &out.baseStr } },
+		{ "baseMag", KeyInfo { &out.baseMag } },
+		{ "baseDex", KeyInfo { &out.baseDex } },
+		{ "baseVit", KeyInfo { &out.baseVit } },
+		{ "maxStr", KeyInfo { &out.maxStr } },
+		{ "maxMag", KeyInfo { &out.maxMag } },
+		{ "maxDex", KeyInfo { &out.maxDex } },
+		{ "maxVit", KeyInfo { &out.maxVit } },
+		{ "blockBonus", KeyInfo { &out.blockBonus } },
+		{ "adjLife", KeyInfo { &out.adjLife } },
+		{ "adjMana", KeyInfo { &out.adjMana } },
+		{ "lvlLife", KeyInfo { &out.lvlLife } },
+		{ "lvlMana", KeyInfo { &out.lvlMana } },
+		{ "chrLife", KeyInfo { &out.chrLife } },
+		{ "chrMana", KeyInfo { &out.chrMana } },
+		{ "itmLife", KeyInfo { &out.itmLife } },
+		{ "itmMana", KeyInfo { &out.itmMana } },
+	};
+	for (DataFileRecord record : dataFile) {
+		FieldIterator fieldIt = record.begin();
+		const FieldIterator endField = record.end();
+		if (fieldIt == endField) {

```

Comment:
This comparison will never be true, `fieldIt` will always contain at least one (potentially zero length) field so will never compare equal to the sentinel until an increment is attempted. Even if you perform a value access this isn't enough to cause the iterator to reach the sentinel value, it has to be incremented.

Reply:
I see, makes sense, removed